patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch

```diff
@@ -8,43 +8,43 @@
        }
  
        iprofiler.func_76319_b();
-@@ -398,6 +399,7 @@
-             if (this.field_219253_g.remove(p_219212_1_, p_219212_3_) && p_219185_5_ != null) {
-                if (p_219185_5_ instanceof Chunk) {
-                   ((Chunk)p_219185_5_).func_177417_c(false);
-+                  net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Unload((Chunk)p_219185_5_));
-                }
- 
-                this.func_219229_a(p_219185_5_);
-@@ -475,6 +477,7 @@
+@@ -475,6 +476,7 @@
                 if (flag) {
                    IChunk ichunk = ChunkSerializer.func_222656_a(this.field_219255_i, this.field_219269_w, this.field_219260_n, p_223172_1_, compoundnbt);
                    ichunk.func_177432_b(this.field_219255_i.func_82737_E());
-+                  net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(ichunk));
++                  net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkDataEvent.Load(ichunk, compoundnbt));
                    return Either.left(ichunk);
                 }
  
-@@ -584,6 +587,7 @@
+@@ -584,6 +586,7 @@
                 if (list != null) {
                    list.forEach(chunk::func_76622_b);
                 }
 +               net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(chunk));
              }
  
              return chunk;
-@@ -668,6 +672,7 @@
+@@ -668,6 +671,7 @@
              }
  
              CompoundNBT compoundnbt1 = ChunkSerializer.func_222645_a(this.field_219255_i, p_219229_1_);
 +            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkDataEvent.Save(p_219229_1_, compoundnbt1));
              this.func_219100_a(chunkpos, compoundnbt1);
              return true;
           } catch (Exception exception) {
-@@ -804,6 +809,7 @@
-       for(int l = i - this.field_219247_A; l <= i + this.field_219247_A; ++l) {
-          for(int k = j - this.field_219247_A; k <= j + this.field_219247_A; ++k) {
-             ChunkPos chunkpos = new ChunkPos(l, k);
-+            net.minecraftforge.event.ForgeEventFactory.fireChunkWatch(p_219234_2_, p_219234_1_, chunkpos, field_219255_i);
-             this.func_219199_a(p_219234_1_, chunkpos, new IPacket[2], !p_219234_2_, p_219234_2_);
+@@ -714,6 +718,7 @@
+ 
+          if (!p_219199_5_ && p_219199_4_) {
+             p_219199_1_.func_213845_a(p_219199_2_);
++            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkWatchEvent.UnWatch(p_219199_1_, p_219199_2_, field_219255_i));
+          }
+ 
+       }
+@@ -1031,6 +1036,7 @@
           }
        }
+ 
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkWatchEvent.Watch(p_219180_1_, p_219180_3_.func_76632_l(), field_219255_i));
```

Comment:
This is the wrong place, this has threading issues to consider. It should probably be moved into setChunkLoadedAtClient